gpiozero/pins/data.py

```diff
@@ -298,7 +489,144 @@ class PinInfo(namedtuple('PinInfo', (
+        the header (1-based)
+
+    .. attribute:: col
+
+        An integer indicating in which column the pin is physically located
+        in the header (1-based)
+    """
+
+
+class HeaderInfo(namedtuple('HeaderInfo', (
+    'name',
+    'rows',
+    'columns',
+    'pins',
+    ))):
+    """
+    This class is a :func:`~collections.namedtuple` derivative used to
+    represent information about a pin header on a board. The object can be used
+    in a format string with various custom specifications::
+
+        from gpiozero import *
+
+        print('{0:full}'.format(pi_info().headers['P1']))
+        print('{0:col2}'.format(pi_info().headers['P1']))
+        print('{0:row1}'.format(pi_info().headers['P1']))
+
+    The following attributes are defined:
+
+    .. automethod:: pprint
+
+    .. attribute:: name
+
+        The name of the header, typically as it appears silk-screened on the
+        board (e.g. "P1").
+
+    .. attribute:: rows
+
+        The number of rows on the header.
+
+    .. attribute:: columns
+
+        The number of columns on the header.
+
+    .. attribute:: pins
+
+        A dictionary mapping physical pin numbers to :class:`PinInfo` tuples.
     """
+    def _func_style(self, function, style):
+        if function == V5:
+            return style('bold red')
+        elif function in (V3_3, V1_8):
+            return style('bold cyan')
+        elif function in (GND, NC):
+            return style('bold black')
+        elif function.startswith('GPIO') and function[4:].isdigit():
+            return style('bold green')
+        else:
+            return style('yellow')
+
+    def _format_full(self, style):
+        Cell = namedtuple('Cell', ('content', 'align', 'style'))
+
+        pin_digits = len(str(self.rows * self.columns))
+        lines = []
+        for row in range(self.rows):
+            line = []
+            for col in range(self.columns):
+                pin = (row * self.columns) + col + 1
+                try:
+                    pin = self.pins[pin]
+                    cells = [
+                        Cell(pin.function, '><'[col % 2], self._func_style(pin.function, style)),
+                        Cell('(%d)' % pin.number, '><'[col % 2], ''),
+                        ]
+                    if col % 2:
+                        cells = reversed(cells)
+                    line.extend(cells)
+                except KeyError:
+                    line.append('')
+            lines.append(line)
+        cols = list(zip(*lines))
+        col_lens = [max(len(cell.content) for cell in col) for col in cols]
+        lines = [
+            ' '.join(
+                '{cell.style}{cell.content:{cell.align}{width}s}{style:reset}'.format(
+                    cell=cell, width=width, style=style)
+                for cell, width, align in zip(line, col_lens, cycle('><')))
+            for line in lines
+            ]
+        return '\n'.join(lines)
+
+    def _format_row(self, row, style):
```

Comment:
It's just occurred to me that both _format_row and _format_col could possibly share a _format_pin function? Not important though.